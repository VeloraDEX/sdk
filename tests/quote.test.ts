import * as dotenv from 'dotenv';
import fetch from 'isomorphic-unfetch';
import {
  constructFetchFetcher,
  constructPartialSDK,
  constructGetQuote,
  isFetcherError,
} from '../src';

import { assert } from 'ts-essentials';

dotenv.config();

jest.setTimeout(30 * 1000);

const ETH = '0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE';
const USDC = '0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48';

const chainId = 1;

const fetchFetcher = constructFetchFetcher(fetch);

describe('Quote:methods', () => {
  const quoteSDK = constructPartialSDK(
    {
      chainId,
      fetcher: fetchFetcher,
      apiURL: process.env.API_URL,
    },
    constructGetQuote
  );

  const amount = '100000000000'; // 100000 USDC,

  test('Get Quote for delta', async () => {
    const quote = await quoteSDK.getQuote({
      srcToken: USDC,
      destToken: ETH,
      amount,
      srcDecimals: 6,
      destDecimals: 18,
      mode: 'delta',
      side: 'SELL',
    });

    expect('delta' in quote).toBeTruthy();
    assert('delta' in quote, 'Delta price not found in Quote');

    const staticDeltaPrice: typeof quote.delta = {
      ...quote.delta,
      hmac: 'dynamic_string',
      destAmount: 'dynamic_number',
      destAmountBeforeFee: 'dynamic_number',
      srcUSD: 'dynamic_number',
      destUSD: 'dynamic_number',
      destUSDBeforeFee: 'dynamic_number',
      gasCost: 'dynamic_number',
      gasCostBeforeFee: 'dynamic_number',
      gasCostUSD: 'dynamic_number',
      gasCostUSDBeforeFee: 'dynamic_number',
    };

    expect(staticDeltaPrice).toMatchInlineSnapshot(`
      {
        "destAmount": "dynamic_number",
        "destAmountBeforeFee": "dynamic_number",
        "destToken": "0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee",
        "destUSD": "dynamic_number",
        "destUSDBeforeFee": "dynamic_number",
        "gasCost": "dynamic_number",
        "gasCostBeforeFee": "dynamic_number",
        "gasCostUSD": "dynamic_number",
        "gasCostUSDBeforeFee": "dynamic_number",
        "hmac": "dynamic_string",
        "partner": "anon",
        "partnerFee": 0,
        "srcAmount": "100000000000",
        "srcToken": "0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48",
        "srcUSD": "dynamic_number",
      }
    `);
  });

  test('Fail Quote for delta for small amounts', async () => {
    const quotePromise = quoteSDK.getQuote({
      srcToken: USDC,
      destToken: ETH,
      amount: (+amount / 1e6).toFixed(0),
      srcDecimals: 6,
      destDecimals: 18,
      mode: 'delta',
      side: 'SELL',
    });

    await expect(quotePromise).rejects.toThrowErrorMatchingInlineSnapshot(
      `"PricingError"`
    );

    const error = await quotePromise.catch((e) => e);

    assert(isFetcherError(error), 'Error should be a FetchError');
    const { details, errorType } = error.response?.data;

    expect({ details, errorType }).toMatchInlineSnapshot(`
      {
        "details": "Error getting price data",
        "errorType": "PricingError",
      }
    `);
  });

  test('Fail to Get Quote for delta with Native Token', async () => {
    const quotePromise = quoteSDK.getQuote({
      srcToken: ETH,
      destToken: USDC,
      amount,
      srcDecimals: 18,
      destDecimals: 6,
      mode: 'delta',
      side: 'SELL',
    });

    await expect(quotePromise).rejects.toThrowErrorMatchingInlineSnapshot(
      `"SourceEth"`
    );

    const error = await quotePromise.catch((e) => e);

    assert(isFetcherError(error), 'Error should be a FetchError');
    const { details, errorType } = error.response?.data;

    expect({ details, errorType }).toMatchInlineSnapshot(`
      {
        "details": "ETH as source token is not supported",
        "errorType": "SourceEth",
      }
    `);
  });

  test('Fail to Get Quote for delta for BUY', async () => {
    const quotePromise = quoteSDK.getQuote({
      srcToken: USDC,
      destToken: ETH,
      amount,
      srcDecimals: 6,
      destDecimals: 18,
      mode: 'delta',
      side: 'BUY',
    });

    await expect(quotePromise).rejects.toThrowErrorMatchingInlineSnapshot(
      `"UnsupportedSide"`
    );

    const error = await quotePromise.catch((e) => e);

    assert(isFetcherError(error), 'Error should be a FetchError');
    const { details, errorType } = error.response?.data;

    expect({ details, errorType }).toMatchInlineSnapshot(`
      {
        "details": "BUY is not supported",
        "errorType": "UnsupportedSide",
      }
    `);
  });

  test('Get Quote for market', async () => {
    const quote = await quoteSDK.getQuote({
      srcToken: USDC,
      destToken: ETH,
      amount,
      srcDecimals: 6,
      destDecimals: 18,
      mode: 'market',
      side: 'SELL',
    });

    expect(quote.market).toBeDefined();
    const priceRoute = quote.market;

    const bestRouteStable = priceRoute.bestRoute
      .map((b) => ({
        ...b,
        percentage: 'dynamic_number',
        percent: 'dynamic_number',
        swaps: 'dynamic_array',
      }))
      .slice(0, 1); // dynamic length, grab first only

    const priceRouteStable = {
      ...priceRoute,
      gasCost: 'dynamic_number',
      gasCostUSD: 'dynamic_number',
      hmac: 'dynamic_string',
      destAmount: 'dynamic_number',
      blockNumber: 'dynamic_number',
      srcUSD: 'dynamic_number',
      destUSD: 'dynamic_number',
      contractMethod: 'dynamic_string',
      bestRoute: bestRouteStable,
    };

    expect(priceRouteStable).toMatchSnapshot();
  });

  test('Get Delta Quote for all', async () => {
    const quote = await quoteSDK.getQuote({
      srcToken: USDC,
      destToken: ETH,
      amount,
      srcDecimals: 6,
      destDecimals: 18,
      mode: 'all',
      side: 'SELL',
    });

    expect('delta' in quote).toBeTruthy();
    assert('delta' in quote, 'Delta price not found in Quote');

    const staticDeltaPrice: typeof quote.delta = {
      ...quote.delta,
      hmac: 'dynamic_string',
      destAmount: 'dynamic_number',
      destAmountBeforeFee: 'dynamic_number',
      srcUSD: 'dynamic_number',
      destUSD: 'dynamic_number',
      destUSDBeforeFee: 'dynamic_number',
      gasCost: 'dynamic_number',
      gasCostBeforeFee: 'dynamic_number',
      gasCostUSD: 'dynamic_number',
      gasCostUSDBeforeFee: 'dynamic_number',
    };

    expect(staticDeltaPrice).toMatchInlineSnapshot(`
      {
        "destAmount": "dynamic_number",
        "destAmountBeforeFee": "dynamic_number",
        "destToken": "0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee",
        "destUSD": "dynamic_number",
        "destUSDBeforeFee": "dynamic_number",
        "gasCost": "dynamic_number",
        "gasCostBeforeFee": "dynamic_number",
        "gasCostUSD": "dynamic_number",
        "gasCostUSDBeforeFee": "dynamic_number",
        "hmac": "dynamic_string",
        "partner": "anon",
        "partnerFee": 0,
        "srcAmount": "100000000000",
        "srcToken": "0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48",
        "srcUSD": "dynamic_number",
      }
    `);
  });

  test('Get Fallback Market Quote for all', async () => {
    const quote = await quoteSDK.getQuote({
      srcToken: USDC,
      destToken: ETH,
      amount: (1e6).toString(),
      srcDecimals: 6,
      destDecimals: 18,
      mode: 'all',
      side: 'SELL',
    });

    assert(!('delta' in quote), 'Delta price not found in quote');

    expect(quote.fallbackReason).toMatchInlineSnapshot(`
      {
        "details": "Gas cost exceeds trade amount",
        "errorType": "GasCostExceedsTradeAmount",
      }
    `);

    const priceRoute = quote.market;

    const bestRouteStable = priceRoute.bestRoute.map((b) => ({
      ...b,
      swaps: b.swaps.map((s) => ({
        ...s,
        swapExchanges: s.swapExchanges.map((se) => ({
          ...se,
          exchange: 'dynamic_string',
          destAmount: 'dynamic_number',
          data: 'largerly dynamic object',
          poolAddresses: 'dynamic_array',
        })),
      })),
    }));

    const priceRouteStable = {
      ...priceRoute,
      gasCost: 'dynamic_number',
      gasCostUSD: 'dynamic_number',
      hmac: 'dynamic_string',
      destAmount: 'dynamic_number',
      blockNumber: 'dynamic_number',
      srcUSD: 'dynamic_number',
      destUSD: 'dynamic_number',
      bestRoute: bestRouteStable,
    };

    expect(priceRouteStable).toMatchSnapshot();
  });
});
