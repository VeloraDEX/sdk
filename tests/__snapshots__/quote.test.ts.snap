// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quote:methods Get Fallback Market Quote for all 2`] = `
{
  "bestRoute": [
    {
      "percent": 100,
      "swaps": [
        {
          "destDecimals": 18,
          "destToken": "0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee",
          "srcDecimals": 6,
          "srcToken": "0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48",
          "swapExchanges": [
            {
              "data": "largerly dynamic object",
              "destAmount": "dynamic_number",
              "exchange": "dynamic_string",
              "percent": 100,
              "poolAddresses": "dynamic_array",
              "srcAmount": "1000000",
            },
          ],
        },
      ],
    },
  ],
  "blockNumber": "dynamic_number",
  "contractAddress": "0x6a000f20005980200259b80c5102003040001068",
  "contractMethod": "swapExactAmountIn",
  "destAmount": "dynamic_number",
  "destDecimals": 18,
  "destToken": "0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee",
  "destUSD": "dynamic_number",
  "gasCost": "dynamic_number",
  "gasCostUSD": "dynamic_number",
  "hmac": "dynamic_string",
  "maxImpact": 100,
  "maxImpactReached": false,
  "network": 1,
  "partner": "anon",
  "partnerFee": 0,
  "side": "SELL",
  "srcAmount": "1000000",
  "srcDecimals": 6,
  "srcToken": "0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48",
  "srcUSD": "dynamic_number",
  "tokenTransferProxy": "0x6a000f20005980200259b80c5102003040001068",
  "version": "6.2",
}
`;

exports[`Quote:methods Get Quote for market 1`] = `
{
  "bestRoute": [
    {
      "percent": "dynamic_number",
      "percentage": "dynamic_number",
      "swaps": "dynamic_array",
    },
  ],
  "blockNumber": "dynamic_number",
  "contractAddress": "0x6a000f20005980200259b80c5102003040001068",
  "contractMethod": "dynamic_string",
  "destAmount": "dynamic_number",
  "destDecimals": 18,
  "destToken": "0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee",
  "destUSD": "dynamic_number",
  "gasCost": "dynamic_number",
  "gasCostUSD": "dynamic_number",
  "hmac": "dynamic_string",
  "maxImpactReached": false,
  "network": 1,
  "partner": "anon",
  "partnerFee": 0,
  "side": "SELL",
  "srcAmount": "100000000000",
  "srcDecimals": 6,
  "srcToken": "0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48",
  "srcUSD": "dynamic_number",
  "tokenTransferProxy": "0x6a000f20005980200259b80c5102003040001068",
  "version": "6.2",
}
`;
